README.md
## Create a backend system for dating APP using Python and Django

## Project summary

The world needs another dating app! But it needs a twist...more like a flip actually! On Upside Town everyones photos are upside down. Your task is to write an API for a tinder-like dating app called Upside Town, that allows people to upload a photo, which then gets flipped upside down on the server, view others flipped photos, match them, and share each others contact information if the match is two-sided.

## Proposer architecture for production system. (Not implemented currently)

![design_document](images/Updown_Town_design_document.png?raw=true)

### Tasks

- Implement assignment using:
  - Language: **Python**
  - Framework: **Django**

- Create 3 API endpoints `/create`, `/list`, and `/match`
  - the `/create` endpoint should allow the user to upload:
    - A photo (limit size to 10M)
      - Flip the photo upside down, then store it
    - Email
    - Gender (m / f)
    - Full name
    - Geo location
      - Geo location should be coordinates (longitude and latitude)

  - the `/match` endpoint should return true if both sides match, and false otherwise:
    - Post t/f
    - Return t/f if both users had their input and decided if to match or not
    - When returning true, send the user the matched persons email

  - the `/list` API should be filterable and paginatable
    - Filter: distance e.g. `/list?gender[eq]=f&distance[lte]=100`
      - You can use the first formula from [this Wikipedia article](https://en.wikipedia.org/wiki/ Great-circle_distance) to calculate distance. Don't forget, you'll need to convert degrees to radians.
    - Pagination: e.g. `/list?page=1`

- Store the data in memory or using sqlite

## Instructions

* Clone the repo to your local machine
  
* Create .env file based on .env_template with all credentials.
  
* Install docker and  run below commands :
  
      docker-compose build
  
      docker-compose up 

* Wait a few minutes for the services to start up.  
  
      docker-compose run web sh start.sh
  
* Wait a few minutes for the services to start up. You can view logs outputs using:
    
      docker-compose -f docker-compose.yml logs -f

## admin UI : 
 * http://localhost:8000/
 
 ![adminui](images/add_profile_at_admin_panel.png?raw=true)

## Rest API: 
 * http://localhost:8000
 
 ![list_api_request](images/list_rest_api.png?raw=true)

 ![create_api_request](images/create_profile_post_api.png?raw=true)

## Print Postgres DB queries in logs for deugging

    docker logs -f postgresql

## Clean up
If you don't have any other docker containers running, you can shut down the ones for this project with the following command:

    docker stop $(docker ps -aq)

Optionally, you can clean up docker images downloaded locally by rinning:

    docker system prune

## TODO: To make it runnable in to production we can add below feature.
  1: CI/CD using git/Jenkins implementation

  2: Unit test cases using pytest module.

  3: Feature test cases using behave framewwork

  4: JWT token based authentication/2-factor authentication to make it secure.

  5: Deployent using Apache

CHANGELOG.md
## 1.0.2 (2021-05-25)

- Merge pull request #1 from abhishek158005/development
- resolved conflicts
- added .git ignore and images for read me file
- Code refactor in docker and docker-compose.yml
- Added system design image
- Added system design document in Readme file
- modified model
- changes for model
- added new model for profile matching
- initial commit with create api


profiles/migrations/0001_initial.py
# Generated by Django 3.2.3 on 2021-05-25 02:49

from django.conf import settings
import django.contrib.gis.db.models.fields
from django.db import migrations, models
import django.db.models.deletion
import django.utils.timezone


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='Profile',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('full_name', models.CharField(max_length=100)),
                ('location', django.contrib.gis.db.models.fields.PointField(srid=4326)),
                ('email', models.CharField(max_length=100)),
                ('gender', models.IntegerField(choices=[(0, 'male'), (1, 'female'), (2, 'not specified')])),
                ('profile_photo', models.ImageField(upload_to='profile_images')),
                ('flipped_profile_photo', models.ImageField(upload_to='flipped_profile_images')),
                ('is_active', models.BooleanField(default=True)),
                ('created_at', models.DateTimeField(auto_now=True)),
                ('updated_at', models.DateTimeField(default=django.utils.timezone.now)),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='user', to=settings.AUTH_USER_MODEL)),
            ],
            options={
                'db_table': 'profiles',
            },
        ),
        migrations.CreateModel(
            name='ProfileMatch',
            fields=[
                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('match_status', models.CharField(choices=[('Pending', 'pending'), ('Rejected', 'rejected'), ('Matched', 'matched')], max_length=100)),
                ('created_at', models.DateTimeField(auto_now=True)),
                ('updated_at', models.DateTimeField(default=django.utils.timezone.now)),
                ('match_profile', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='match_profile', to='profiles.profile')),
                ('requested_profile', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='requested_profile', to='profiles.profile')),
            ],
            options={
                'db_table': 'profile_matches',
            },
        ),
    ]
